### ISSUE TITLE
Empty progress bars never finish

### ISSUE THREAD

**Issue by kormoc** [vbauerster/mpb#7]
When using the ProxyReader with empty files, progress bars never complete and end up locking up the program on exit. This also seems to happen with empty normal progress bars

**Comment by vbauerster** [comment]
Well that's intended behaviour. It happens when `total <= 0` in `b := p.AddBar(total)`. **mpb** treats such case as total is unknown, and a user suppose to use [b.Completed()](https://godoc.org/github.com/vbauerster/mpb#Bar.Completed) method after work is done, see [this](https://gist.github.com/vbauerster/24290cd1038c76fc585f9ccdd6ade41c) gist for example.
Why not just return if `total <= 0`?
Suppose one want to download a file over http. Some http servers may report size 0, but `resp.Body` is not empty. Usually this happens when http server doesn't support resume. So rather than return silently, **mpb** fallbacks to [simple spinner](https://github.com/vbauerster/mpb/blob/6013267b5c3d60409c5c34800d4832b813d6178d/bar.go#L275-L277) bar, and expects user to call [b.Completed()](https://godoc.org/github.com/vbauerster/mpb#Bar.Completed) method, after job done.

**Comment by vbauerster** [comment]
I've implemented a fix at [issue7fix](https://github.com/vbauerster/mpb/tree/issue7fix) branch, which will call `b.Completed()` implicitly.
Can you please checkout this branch, and check if it works for you?

**Comment by kormoc** [comment]
I can confirm that this fixes the exit freeze, however it does display things oddly

```
foobar                                      0b /  0b [/] -9223372036854775808 %
```

Thanks for the quick fix though!

**Comment by vbauerster** [comment]
Can you show your code? How do you initialise a bar?

**Comment by kormoc** [comment]
Certainly, the code is here:

https://github.com/kormoc/bitrot_scanner/blob/master/checksumHelpers.go#L53

**Comment by vbauerster** [comment]
Yes negative percentage is odd looking. I've fixed that, and merged to master. You can checkout.

**Comment by vbauerster** [comment]
Looking at your code 
```go
PrependCounters("%3s / %3s", mpb.UnitBytes, 18, mpb.DwidthSync|mpb.DextraSpace)
```
 I assume you use multiple bars. For single bar, option `mpb.DwidthSync` has no effect.

**Comment by vbauerster** [comment]
As for simple spinner, `[/]` part, it cycles through  a [set](https://github.com/vbauerster/mpb/blob/6d8f22f09781294b1540b0b6950c75337481c985/bar.go#L532) of `-\|/` characters. Maybe I should let user to decide whether he/she wants this spinner for such edge case (total <= 0)? It is on by default, what do you think?

**Comment by kormoc** [comment]
Yes, we (optionally) use multiple bars. I don't mind the spinner in this case. Thanks much for the quick fix!

**Comment by vbauerster** [comment]
You're welcome, thanks for reporting!

**Comment by falzm** [comment]
Hi, sorry to dig this issue out of its grave but it looks like this "problem" is still present with current versions (tested with v4.9.1). The code has much changed since this issue's fix commit, I tried to track down the code but can't seem to find it in the current code base so I wonder if it may have been forgotten during a refactoring maybe?

Our actual implementation is here: https://github.com/exoscale/cli/blob/master/cmd/sos_upload.go#L170-L194
If we pass the raw file reader to `PutObjectWithContext()` method it works fine with empty files, but get stuck if we pass the `bar.ProxyReader(f)` reader.

**Comment by vbauerster** [comment]
Hi, you're right, `Read` method of `proxyReader` never checked EOF. Should be fixed with [this commit](https://github.com/vbauerster/mpb/commit/08e821e86ad39292506c19fb3a25a94f61615272). Can you pls check, and confirm if it's fixed?
I'll bump tag after your confirmation.

**Comment by falzm** [comment]
@vbauerster tested with this commit: no improvement.

**Comment by vbauerster** [comment]
I've simulated empty file/reader in `io` example, by changing [this line](https://github.com/vbauerster/mpb/blob/53fa35fbcecf837029116ace8ea3a5d6a17e3da4/_examples/io/main.go#L14) to `total := 0` and there is no freeze.

**Comment by vbauerster** [comment]
Are you sure your build has picked up aforementioned commit?

**Comment by falzm** [comment]
Yes:

```console
$ git diff go.mod
diff --git a/go.mod b/go.mod
index efb439f..5a23218 100644
--- a/go.mod
+++ b/go.mod
@@ -23,7 +23,7 @@ require (
        github.com/spf13/cobra v0.0.3
        github.com/spf13/pflag v1.0.3
        github.com/spf13/viper v1.3.1
-       github.com/vbauerster/mpb/v4 v4.8.4
-       golang.org/x/crypto v0.0.0-20190530122614-20be4c3c3ed5
+       github.com/vbauerster/mpb/v4 v4.9.2-0.20190806152146-08e821e86ad3
+       golang.org/x/crypto v0.0.0-20190701094942-4def268fd1a4
        gopkg.in/alecthomas/kingpin.v3-unstable v3.0.0-20180810215634-df19058c872c // indirect
 )

$ git diff vendor/github.com/vbauerster/mpb/v4/proxyreader.go
diff --git a/vendor/github.com/vbauerster/mpb/v4/proxyreader.go b/vendor/github.com/vbauerster/mpb/v4/proxyreader.go
index d2692cc..1e172be 100644
--- a/vendor/github.com/vbauerster/mpb/v4/proxyreader.go
+++ b/vendor/github.com/vbauerster/mpb/v4/proxyreader.go
@@ -18,5 +18,9 @@ func (pr *proxyReader) Read(p []byte) (n int, err error) {
                pr.bar.IncrBy(n, time.Since(pr.iT))
                pr.iT = time.Now()
        }
+       if err == io.EOF {
+               current := pr.bar.Current()
+               pr.bar.SetTotal(current, true)
+       }
        return
 }

$ make
go clean
rm -rf exo contrib/completion manpage
go build -ldflags "-s -w -X main.version=dev -X main.commit=b7f5c7dd64d3c8c08d6994519cb2d76ffabebc39" -mod=vendor -o exo

$ ./exo version
exo dev b7f5c7dd64d3c8c08d6994519cb2d76ffabebc39 (egoscale 0.18.1)

$ touch blah
$ ./exo sos upload marc-test blah
0s 0 % [--------------------------------------------------------------] blah
# ^^^ gets stuck ^^^

**Comment by falzm** [comment]
You can perform the same test on your end by checking out our repository: https://github.com/exoscale/cli

**Comment by vbauerster** [comment]
It asks for config which I don't have. I wonder if `Reader` method is called at all on empty file. You can troubleshoot by putting `fmt.Fprintln(os.Stderr, "read called")` [here](https://github.com/vbauerster/mpb/blob/53fa35fbcecf837029116ace8ea3a5d6a17e3da4/proxyreader.go#L16), then run like `./exo 2> debug.txt`. I recommend using [gohack](https://github.com/rogpeppe/gohack) for that.

**Comment by falzm** [comment]
Tested with your instructions: the `debug.txt` is empty.

**Comment by falzm** [comment]
> Tested with your instructions: the `debug.txt` is empty.

Wait nevermind, I think the debug line has not been taken by the compiler via gohack. Let me try again.

**Comment by falzm** [comment]
This time I'm positive the modified file is actually used because I got compilation error complaining about missing `fmt`/`os` package imports.

Sadly, the `debug.txt` remains empty:

```console
$ git diff vendor/github.com/vbauerster/mpb/v4/proxyreader.go
diff --git a/vendor/github.com/vbauerster/mpb/v4/proxyreader.go b/vendor/github.com/vbauerster/mpb/v4/proxyreader.go
index d2692cc..d4b142e 100644
--- a/vendor/github.com/vbauerster/mpb/v4/proxyreader.go
+++ b/vendor/github.com/vbauerster/mpb/v4/proxyreader.go
@@ -1,7 +1,9 @@
 package mpb

 import (
+       "fmt"
        "io"
+       "os"
        "time"
 )

@@ -13,10 +15,15 @@ type proxyReader struct {
 }

 func (pr *proxyReader) Read(p []byte) (n int, err error) {
+       fmt.Fprintln(os.Stderr, "read called")
        n, err = pr.ReadCloser.Read(p)
        if n > 0 {
                pr.bar.IncrBy(n, time.Since(pr.iT))
                pr.iT = time.Now()
        }
+       if err == io.EOF {
+               current := pr.bar.Current()
+               pr.bar.SetTotal(current, true)
+       }
        return
 }

$ make
go clean
rm -rf exo contrib/completion manpage
go build -ldflags "-s -w -X main.version=dev -X main.commit=b7f5c7dd64d3c8c08d6994519cb2d76ffabebc39" -mod=vendor -o exo


$ ./exo sos upload marc-test blah 2>debug.txt
0s 0 % [--------------------------------------------------------------] blah
^C 
$ cat debug.txt
$
```

**Comment by vbauerster** [comment]
Well it means that `minioClient` doesn't handle it properly, however as workaround you may consider checking empty file manually and call `bar.SetTotal(0, true)` afterwards. Long term fix is to report upstream to `minio` lib.

**Comment by falzm** [comment]
Well, indeed your workaround works – thank you for that. I'll go ahead and open an issue at Minio, you can close this one.
